astropy/utils/data.py

```diff
@@ -968,154 +1051,113 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
+            errors[source_url] = e
+        except socket.timeout as e:
+            # this isn't supposed to happen, but occasionally a socket.timeout
+            # gets through.  It's supposed to be caught in urllib and raised
+            # in this way, but for some reason in mysterious circumstances it
+            # doesn't (or didn't in python2?). So we'll just re-raise it here
+            # instead.
+            errors[source_url] = e
+    else:   # No success
+        if not sources:
+            raise KeyError(
+                f"No sources listed and file {remote_url} not in cache! "
+                f"Please include primary URL in sources if you want it to be "
+                f"included as a valid source.")
+        elif len(sources) == 1:
+            raise errors[sources[0]]
         else:
-            local_path = f.name
-            if missing_cache:
-                msg = ('File downloaded to temporary location due to problem '
-                       'with cache directory and will not be cached.')
-                warn(CacheMissingWarning(msg, local_path))
-            if conf.delete_temporary_downloads_at_exit:
-                global _tempfilestodel
-                _tempfilestodel.append(local_path)
-    except urllib.error.URLError as e:
-        if hasattr(e, 'reason') and hasattr(e.reason, 'errno') and e.reason.errno == 8:
-            e.reason.strerror = e.reason.strerror + '. requested URL: ' + remote_url
-            e.reason.args = (e.reason.errno, e.reason.strerror)
-        raise e
-    except socket.timeout as e:
-        # this isn't supposed to happen, but occasionally a socket.timeout gets
-        # through.  It's supposed to be caught in `urrlib2` and raised in this
-        # way, but for some reason in mysterious circumstances it doesn't. So
-        # we'll just re-raise it here instead
-        raise urllib.error.URLError(e)
-
-    return local_path
+            raise urllib.error.URLError(
+                f"Unable to open any source! Exceptions were {errors}") \
+                from errors[sources[0]]
+
+    if cache:
+        try:
+            return import_file_to_cache(url_key, f_name,
+                                        hexdigest=hexdigest,
+                                        remove_original=True)
+        except PermissionError:
+            # Cache is readonly, we can't update it
+            missing_cache = (
+                f"Cache directory appears to be read-only, unable to import "
+                f"downloaded file, providing data in temporary file {f_name} "
+                f"instead.")
+
+    if missing_cache:
+        warn(CacheMissingWarning(missing_cache, f_name))
+    if conf.delete_temporary_downloads_at_exit:
+        global _tempfilestodel
+        _tempfilestodel.append(f_name)
+    return f_name
 
 
 def is_url_in_cache(url_key):
-    """
-    Check if a download from ``url_key`` is in the cache.
+    """Check if a download from ``url_key`` is in the cache.
```

Comment:
With the introduction of `sources`, I feel like this can use some clarification. Feel free to reword if you don't like what I wrote below.
```suggestion
    """Check if a download from ``url_key`` is in the cache.
    Note that you have to provide the main URL, not any of the mirrors you used,
    because this function does not check for contents, only the stored hash.
```

Reply:
I'm not quite sure where to provide a general explanation of the cache. Basically what I want to say is: 

"The cache is effectively a dictionary mapping URLs to files; by default the file contains the contents of the URL that is its key, but in practice these can be obtained from a mirror or imported from the local filesystem. Regardless, each file is regarded as representing the contents of a particular URL, and this URL should be used to look them up or otherwise manipulate them." 

But I'm not sure where to put this that it is hard to miss when reading (say) the docstrings of minor functions like this. It seems like I keep having to re-explain the relationship between key and contents. In a way this is reasonable, because we don't know which docstring a user will check first. Though probably download_file. How about I put an explanation there, then come up with a two-line version I can paste into all the other functions along with a reference to download_file's docstring?

Can I easily check that the docs are rendering okay? I'm never sure how to make cross-references work in sphinx plus napoleon.

Reply:
Well... How about a sub-section in "Downloadable data management" (if it is not there already) and the API doc and link to that section? Something like (badly worded here, pls rephrase):

```
.. note::

    Cache only recognize its key that is the main URL,
    see :ref:`such-and-such-section` for more info.
```
:woman_shrugging:

Reply:
Also, I was just thinking... If the *main* URL itself changes over time, we might end up with multiple copies of the same file, just stored under different URL hashes. Not sure if there is much we can do about it, except to educate the devs/users.

Reply:
> Also, I was just thinking... If the _main_ URL itself changes over time, we might end up with multiple copies of the same file, just stored under different URL hashes. Not sure if there is much we can do about it, except to educate the devs/users.

I think I would treat that as a case of a URL that stopped being used - like any such, the downloaded data corresponding to it will stick around until manually cleared out of the cache. On UNIX one could check the atime of the file, on Windows it'll probably stick around until you have to wipe and reinstall. A user could use cache_contents to browse the files there (including noticing that some URLs shared contents). Down the road, if this proves to be a regular issue, we could enhance these tools to make it easier to detect long-unused files automatically.

Reply:
Yeah, don't worry about that right now. We have a long standing issue that is #1162 . :sweat_smile: